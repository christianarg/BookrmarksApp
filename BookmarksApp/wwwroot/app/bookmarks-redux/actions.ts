import { BookmarkModel, TagModelState, AddOrEditTagResult, EditBookmark } from "./bookmark-model";

export const ADD_BOOKMARK = 'ADD_BOOKMARK';
export const ADD_OR_EDIT_TAG = 'ADD_OR_EDIT_TAG';
export const EDIT_BOOKMARK = 'EDIT_BOOKMARK';
export const SEARCH = 'SEARCH';

export type AddBookmark = {
    type: typeof ADD_BOOKMARK;
    tagName: string;
    bookmarkModel: BookmarkModel;
}

export function addBookmark(bookmarkModel: BookmarkModel, tagName: string): AddBookmark {
    return {
        type: ADD_BOOKMARK,
        bookmarkModel: bookmarkModel, 
        tagName: tagName
    }
}

//export type AddOrEditTag = {
//    type: typeof ADD_OR_EDIT_TAG;
//    addOrEditTagResult: AddOrEditTagResult;
//    parentTag: TagModel;
//}

//export function addOrEditTag(addOrEditTagResult: AddOrEditTagResult, parentTag: TagModel): AddOrEditTag  {
//    return { ...args, type: ADD_OR_EDIT_TAG }
//}

//export type EditBookmark = {
//    tag: TagModel, editedBookmark: EditBookmark
//}

//export function editBookmark(args: EditBookmark): EditBookmark {
//    return { ...args, type: EDIT_BOOKMARK }
//}

export function search(searchValue: string) {
    return {
        type: SEARCH,
        searchValue: searchValue
    }
}


export type BookmarkActionTypes =
    AddBookmark;
    //|
    //EditBookmark |
    //AddOrEditTag;

