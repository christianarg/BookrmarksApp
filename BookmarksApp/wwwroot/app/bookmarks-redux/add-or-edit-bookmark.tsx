import * as React from 'react';
import { BookmarkModel, EditBookmark } from './bookmark-model';

type AddOrEditBookmarkProps = {
    bookmarkToEdit?: BookmarkModel;
    onAddOrEdit: (newBookmark: EditBookmark) => void;
};
type AddOrEditBookmarkState = {
    isFormVisible: boolean;
    name: string;
    url: string;
};
export class AddOrEditBookmark extends React.Component<AddOrEditBookmarkProps, AddOrEditBookmarkState> {
    constructor(props) {
        super(props);
        const bookmarkToEdit = this.props.bookmarkToEdit;
        if (bookmarkToEdit) {
            this.state = { isFormVisible: false, name: bookmarkToEdit.name, url: bookmarkToEdit.url };
        }
        else {
            this.state = { isFormVisible: false, name: "", url: "" };
        }
    }
    hasValue() {
        const { name, url } = this.state;
        return (name && url);
    }
    handleSubmit = (evt: React.FormEvent) => {
        evt.preventDefault();
        const { name, url } = this.state;
        const bookmarkToEdit = this.props.bookmarkToEdit;
        if (name && url) {
            this.props.onAddOrEdit({ name: name, url: url, oldName: bookmarkToEdit && bookmarkToEdit.name });
            this.setState({ isFormVisible: false, url: "", name: "" });
        }
    };
    handleNameChange = (evt: React.ChangeEvent<HTMLInputElement>) => {
        this.setState({ name: evt.target.value });
    };
    handleUrlChange = (evt: React.ChangeEvent<HTMLInputElement>) => {
        this.setState({ url: evt.target.value });
    };
    toggleShow = () => {
        this.setState({ isFormVisible: !this.state.isFormVisible });
    };
    buttonStyle(): React.CSSProperties {
        if (!this.hasValue()) {
            return { cursor: 'not-allowed' };
        }
        return null;
    }
    render() {
        let bookmarkToEdit = this.props.bookmarkToEdit;
        const idEdit = bookmarkToEdit != null;
        const addOrEditToggleButtonText = bookmarkToEdit ? '(Edit Bookmark)' : '(Add Bookmark)';
        const addOrEdditAcceptButtonText = bookmarkToEdit ? 'Edit' : 'Add';
        const buttonStyle = { display: 'inline-block', textDecoration: 'underline', cursor: 'pointer' };
        if (this.state.isFormVisible) {
            return (<div>
                <div style={buttonStyle} onClick={this.toggleShow}>Close</div>
                <form onSubmit={this.handleSubmit}>
                    <div>Name: <input type="text" value={this.state.name} onChange={this.handleNameChange} placeholder="Bookmark name..." /></div>
                    <div>Url:  <input type="text" value={this.state.url} onChange={this.handleUrlChange} placeholder="url..." /></div>
                    <button value="Add" style={this.buttonStyle()}>{addOrEdditAcceptButtonText}</button>
                </form>
            </div>);
        }
        return <div style={buttonStyle} onClick={this.toggleShow}>{addOrEditToggleButtonText}</div>;
    }
}
