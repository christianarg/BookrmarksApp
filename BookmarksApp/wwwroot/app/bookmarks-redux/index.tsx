import * as React from 'react';
import { BookmarkModel, TagModel, AddOrEditTagResult, EditBookmark } from "./bookmark-model";
import { replaceTag, filterTags } from './common';
import { TagSearch } from './TagSearchProps';
import { Tags } from './Tags';
import { AddOrEditTag } from './AddTagProps';

export type TagsRootState = {
    tags: TagModel[];
    searachText: string;
}

export class TagsRoot extends React.Component<{}, TagsRootState> {
    state: TagsRootState = { tags: null, searachText: '' };
    componentDidMount() {
        this.setState({ tags: sampleBookrmarks.slice() });
    }
    handleSearch = (searchValue: string) => {
        const tags = this.state.tags.slice();
        filterTags(tags, searchValue);
        this.setState({ tags: tags, searachText: searchValue });
    };
    handleAddBookmark = (tag: TagModel, bookmarkModel: BookmarkModel) => {
        let tags = this.state.tags.slice();
        tag.bookmarks.push(bookmarkModel);
        tags = replaceTag(tags, tag);
        this.setState({ tags: tags });
    };
    handleAddOrEditTag = (addOrEditTagResult: AddOrEditTagResult, parentTag: TagModel) => {
        let tags = this.state.tags.slice();
        if (parentTag) {
            if (parentTag.subTags == null)
                parentTag.subTags = [];
            if (addOrEditTagResult.oldName) {
                // replace
                parentTag.subTags = parentTag.subTags.map(x => x.name == addOrEditTagResult.oldName ? addOrEditTagResult : x);
            }
            else {
                // new
                parentTag.subTags.push(addOrEditTagResult);
            }
            replaceTag(tags, { ...parentTag });
        }
        else {
            // sin parent
            if (addOrEditTagResult.oldName) {
                // replace
                tags = tags.map(x => x.name == addOrEditTagResult.oldName ? addOrEditTagResult : x);
            }
            else {
                // new
                tags.push(addOrEditTagResult);
            }
        }
        this.setState({ tags: tags });
    };
    handleEditBookmark = (tag: TagModel, editedBookmark: EditBookmark) => {
        let tags = this.state.tags.slice();
        // replace bookmark
        tag.bookmarks = tag.bookmarks.map(x => x.name == editedBookmark.oldName ? editedBookmark : x);
        tags = replaceTag(tags, tag);
        this.setState({ tags: tags });
    };
    render() {
        if (this.state.tags) {
            return (<div>
                <TagSearch searachText={this.state.searachText} onSearchChange={this.handleSearch} />
                <Tags tags={this.state.tags} parentTag={null} onEditBookmark={this.handleEditBookmark} onAddBookmark={this.handleAddBookmark} onAddTag={this.handleAddOrEditTag} />
                <AddOrEditTag isRoot={true} onAddOrEdit={(newTag) => this.handleAddOrEditTag(newTag, null)} />
            </div>);
        }
        return null;
    }
}



export const sampleBookrmarks: TagModel[] = [
    {
        name: '.Net',
        bookmarks: [
            { name: 'download', url: 'https://dotnet.microsoft.com/download' }
        ],
        subTags: [
            {
                name: 'Asp.net',
                bookmarks: [
                    { name: 'asp.net', url: 'https://dotnet.microsoft.com/apps/aspnet' },
                    { name: 'asp.net core', url: 'https://docs.microsoft.com/es-es/aspnet/core/?view=aspnetcore-2.2' },
                ],
            }
        ]
    },
    {
        name: 'React',
        bookmarks: [
            { name: 'react docs', url: 'https://reactjs.org/docs/getting-started.html' }
        ]
    }
]