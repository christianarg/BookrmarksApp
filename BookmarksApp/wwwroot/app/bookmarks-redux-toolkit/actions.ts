import { BookmarkModel, TagModelState, AddOrEditTagResult, EditBookmark } from "./bookmarks-redux";

export const ADD_BOOKMARK = 'ADD_BOOKMARK';
export const ADD_OR_EDIT_TAG = 'ADD_OR_EDIT_TAG';
export const EDIT_BOOKMARK = 'EDIT_BOOKMARK';
export const SEARCH = 'SEARCH';

export type AddBookmark = {
    type: typeof ADD_BOOKMARK;
    tagName: string;
    bookmarkModel: BookmarkModel;
}

export function addBookmark(bookmarkModel: BookmarkModel, tagName: string): AddBookmark {
    return {
        type: ADD_BOOKMARK,
        bookmarkModel: bookmarkModel, 
        tagName: tagName
    }
}

export type AddOrEditTag = {
    type: typeof ADD_OR_EDIT_TAG;
    addOrEditTagResult: AddOrEditTagResult;
    parentTagName: string;
}


export function addOrEditTag(addOrEditTagResult: AddOrEditTagResult, parentTagName: string): AddOrEditTag  {
    return {
        type: ADD_OR_EDIT_TAG,
        addOrEditTagResult: addOrEditTagResult,
        parentTagName: parentTagName
    }
}

export type Search = {
    type: typeof SEARCH;
    searchValue: string;
}


export function search(searchValue: string): Search {
    return {
        type: SEARCH,
        searchValue: searchValue
    }
}


export type BookmarkActionTypes =
    AddBookmark | AddOrEditTag | Search;