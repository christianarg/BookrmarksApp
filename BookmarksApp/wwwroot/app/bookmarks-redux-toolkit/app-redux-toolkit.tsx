import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { ConnectedTagsRoot } from "./bookmarks-redux-toolkit";
import { Provider } from 'react-redux'
import { configureStore } from '@reduxjs/toolkit'
import reducer, { recieveState, sampleState, addBookmark} from './bookmark-slice'

const store = configureStore({ reducer: reducer });

export class App extends React.Component {
    componentDidMount() {
        // Estado inicial de pruebas (aqui habría que hacer un fetch)
        store.dispatch(recieveState({ newState: sampleState }));
    }
    render() {
        return (
            <Provider store={store}>
                <ConnectedTagsRoot />
            </Provider>);
    }
}

ReactDOM.render(<App />, document.getElementById('root'));


