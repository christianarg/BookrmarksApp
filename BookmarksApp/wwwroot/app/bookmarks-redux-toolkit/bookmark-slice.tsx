import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { BookmarksAppState, TagModelState, BookmarkModel, EditBookmark } from './bookmarks-redux-toolkit';
import { AddOrEditTagResult } from './bookmarks-redux-toolkit';

const initialState: BookmarksAppState = {
    bookmarks: [],
    tags: [],
    searchValue: ''
};

export const sampleState: BookmarksAppState = {
    tags:
        [
            {
                name: '.Net',
                bookmarks: ['download'],
                subTags: ['Asp.net'],
                isRoot: true
            },
            {
                name: 'Asp.net',
                bookmarks: ['asp.net', 'asp.net core']
            },
            {
                name: 'React',
                bookmarks: ['react docs'],
                isRoot: true
            }
        ],
    bookmarks:
        [
            { name: 'download', url: 'https://dotnet.microsoft.com/download' },
            { name: 'asp.net', url: 'https://dotnet.microsoft.com/apps/aspnet' },
            { name: 'asp.net core', url: 'https://docs.microsoft.com/es-es/aspnet/core/?view=aspnetcore-2.2' },
            { name: 'react docs', url: 'https://reactjs.org/docs/getting-started.html' }
        ],
    searchValue: ''
};

/**Para tests o carga inicial */
export type SetAllState = {
    newState: BookmarksAppState;
}

export type AddBookmark = {
    tagName: string;
    bookmarkModel: BookmarkModel;
}

export type AddOrEditTagParams = {
    addOrEditTagResult: AddOrEditTagResult;
    parentTagName: string;
}

export type Search = {
    searchValue: string;
}

const bookmarksSlice = createSlice({
    name: 'bookmarks',
    initialState: initialState,
    reducers: {
        recieveState(state, action: PayloadAction<SetAllState>) {
            state.bookmarks = action.payload.newState.bookmarks;
            state.tags = action.payload.newState.tags;
        },
        addBookmark(state, action: PayloadAction<AddBookmark>) {
            let { tags, bookmarks } = state;
            const { tagName, bookmarkModel } = action.payload;
            let tag = tags.find(x => x.name == tagName);
            tag.bookmarks.push(bookmarkModel.name);
            bookmarks.push(bookmarkModel);    // añadir bookmark
            //tags = tags.map(x => x.name == tag.name ? tag : x); // reeplazar tag (no hace falta por el immer)

            return state;
        },
        addOrEditTag(state, action: PayloadAction<AddOrEditTagParams>) {
            let tags = state.tags;
            const { addOrEditTagResult, parentTagName } = action.payload;
            let tagToAddOrEdit = state.tags.find(x => x.name == addOrEditTagResult.oldName);
            if (tagToAddOrEdit) {  // edit
                tagToAddOrEdit.name = addOrEditTagResult.name;
                tags = tags.map(x => x.name == addOrEditTagResult.oldName ? tagToAddOrEdit : x);

                // en el padre reemplazar el subtag por el nuevo nombre
                let parentTag = state.tags.find(x => x.name == parentTagName);
                if (parentTag) {
                    parentTag.subTags = parentTag.subTags.map(x => x == addOrEditTagResult.oldName ? tagToAddOrEdit.name : x);
                    tags = tags.map(x => x.name == parentTag.name ? parentTag : x);
                }
            } else {
                tagToAddOrEdit = addOrEditTagResult;
                tags.push(tagToAddOrEdit);
                let parentTag = tags.find(x => x.name == parentTagName);
                if (parentTag) {
                    if (!parentTag.subTags) {
                        parentTag.subTags = [];
                    }
                    parentTag.subTags.push(tagToAddOrEdit.name);
                    tags = tags.map(x => x.name == parentTag.name ? parentTag : x);
                }
                else {
                    tagToAddOrEdit.isRoot = true;
                }
            }
            return state;
        },
        editBookmark(state, action: PayloadAction<EditBookmark>) {
            const { oldName, name } = action.payload
            let { bookmarks, tags } = state;

            // reemplazar el bookmar en el estado global (se podría hacer con un map, estoy probando esta manera altenrativa)
            const bookMarkIndex = bookmarks.findIndex(x => x.name == oldName);
            bookmarks[bookMarkIndex] = action.payload;

            // si cambia el nombre hay que reemplazarlo en el bookmark del tag (si hacen falta id's sino esto se puede liar, o bien que el bookmark tenga el tagname)
            if (name != oldName) {
                const tagWithChangedBookmark = tags.find(x => x.bookmarks.some(bookmarkName => bookmarkName == oldName));
                const bookMarkIndexInTag = tagWithChangedBookmark.bookmarks.findIndex(x => x == oldName);
                tagWithChangedBookmark.bookmarks[bookMarkIndexInTag] = name;
            }
            return state;
        },
        search(state: BookmarksAppState, action: PayloadAction<Search>) {
            const searchText = action.payload.searchValue;
            state.searchValue = searchText;
            filterTags(state, state.tags, state.searchValue);
           
            return state;
        }
    }
});

function hasText(text: string, searchText: string) {
    return text.toLowerCase().includes(searchText.toLowerCase());
}

function getTags(state: BookmarksAppState, tagNames: string[]): TagModelState[] {
    const allTags = state.tags;

    return allTags.filter(x => tagNames.some(tagName => tagName == x.name));
}

function getBookmarks(state: BookmarksAppState, bookmarkNames: string[]): BookmarkModel[] {
    const allBookmarks = state.bookmarks;

    return allBookmarks.filter(x => bookmarkNames.some(bookmarkName => bookmarkName == x.name));
}

export function filterTags(state: BookmarksAppState, tags: TagModelState[], searchText: string) {
    const allTags = state.tags;

    tags.forEach(tag => {
        const bookmarks = getBookmarks(state, tag.bookmarks);

        bookmarks.forEach(bookmark => bookmark.hidden = !hasText(bookmark.name, searchText));
        const anyBookmarkVisible = bookmarks.some(b => !b.hidden);

        const tagNameHasText = hasText(tag.name, searchText);

        const anySubTagsVisible = tag.subTags && getTags(state, tag.subTags).some(t => !t.hidden);
        tag.hidden = !tagNameHasText && !anyBookmarkVisible && !anySubTagsVisible;
        if (tag.subTags) {
            filterTags(state, getTags(state, tag.subTags), searchText);
        }
    });
}

export const {
    addBookmark,
    addOrEditTag,
    editBookmark,
    search,
    recieveState
} = bookmarksSlice.actions;

export default bookmarksSlice.reducer;