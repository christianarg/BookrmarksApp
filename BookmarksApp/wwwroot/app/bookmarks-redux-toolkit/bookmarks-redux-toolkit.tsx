import * as React from 'react';
import { connect, MapDispatchToPropsParam } from "react-redux";
import { Action } from 'redux';
import { addBookmark, addOrEditTag, editBookmark, search, AddBookmark, AddOrEditTagParams } from './bookmark-slice'

// **Models**
export type BookmarksAppState = {
    bookmarks: BookmarkModel[];
    tags: TagModelState[];
}

export type BookmarkModel = {
    name: string;
    url: string;
    hidden?: boolean;
};

export type TagModelState = {
    name: string;
    isRoot?: boolean;
    subTags?: string[];
    bookmarks: string[];
    hidden?: boolean;
}

export type AddOrEditTagResult = TagModelState & { oldName?: string; }

export type EditBookmark = BookmarkModel & { oldName: string };

// **BookmarkComponent**

type BookmarkProps = {
    bookmarks: BookmarkModel[];
    onEdit: (bookrmark: EditBookmark) => void;
};
export function Bookmarks(props: BookmarkProps) {
    const bookmarks = props.bookmarks.map(bookmark => !bookmark.hidden &&
        <li key={bookmark.name} className="bookmark">
            <a href={bookmark.url} target="_blank">{bookmark.name}</a>&nbsp;&nbsp;<AddOrEditBookmark bookmarkToEdit={bookmark} onAddOrEdit={edited => props.onEdit(edited)} />
        </li>);
    return (<ul style={{ listStyleType: 'square' }}>{bookmarks}</ul>);
}

// **AddOrEditBookmarkComponent**

type AddOrEditBookmarkProps = {
    bookmarkToEdit?: BookmarkModel;
    onAddOrEdit: (newBookmark: EditBookmark) => void;
};

type AddOrEditBookmarkState = {
    isFormVisible: boolean;
    name: string;
    url: string;
};

export class AddOrEditBookmark extends React.Component<AddOrEditBookmarkProps, AddOrEditBookmarkState> {

    state: AddOrEditBookmarkState = { isFormVisible: false, name: "", url: "" };

    componentDidMount() {
        this.updateFormWithBookmark();
    }

    componentDidUpdate(prevProps: AddOrEditBookmarkProps) {
        if (prevProps.bookmarkToEdit != this.props.bookmarkToEdit) {
            this.updateFormWithBookmark();
        }
    }

    updateFormWithBookmark() {
        const bookmarkToEdit = this.props.bookmarkToEdit;
        if (bookmarkToEdit) {
            this.setState({ name: bookmarkToEdit.name, url: bookmarkToEdit.url });
        }
        else {
            this.setState({ name: "", url: "" });
        }
    }

    hasValue() {
        const { name, url } = this.state;
        return (name && url);
    }


    handleSubmit = (evt: React.FormEvent) => {
        evt.preventDefault();
        const { name, url } = this.state;
        const bookmarkToEdit = this.props.bookmarkToEdit;
        if (name && url) {
            this.props.onAddOrEdit({ name: name, url: url, oldName: bookmarkToEdit && bookmarkToEdit.name });
            this.setState({ isFormVisible: false, url: "", name: "" });
        }
    };

    handleNameChange = (evt: React.ChangeEvent<HTMLInputElement>) => {
        this.setState({ name: evt.target.value });
    };
    handleUrlChange = (evt: React.ChangeEvent<HTMLInputElement>) => {
        this.setState({ url: evt.target.value });
    };
    toggleShow = () => {
        this.setState({ isFormVisible: !this.state.isFormVisible });
    };

    buttonStyle(): React.CSSProperties {
        if (!this.hasValue()) {
            return { cursor: 'not-allowed' };
        }
        return null;
    }

    render() {
        let bookmarkToEdit = this.props.bookmarkToEdit;
        const idEdit = bookmarkToEdit != null;
        const addOrEditToggleButtonText = bookmarkToEdit ? '(Edit Bookmark)' : '(Add Bookmark)';
        const addOrEdditAcceptButtonText = bookmarkToEdit ? 'Edit' : 'Add';
        const buttonStyle = { display: 'inline-block', textDecoration: 'underline', cursor: 'pointer' };
        if (this.state.isFormVisible) {
            return (<div>
                <div style={buttonStyle} onClick={this.toggleShow}>Close</div>
                <form onSubmit={this.handleSubmit}>
                    <div>Name: <input type="text" value={this.state.name} onChange={this.handleNameChange} placeholder="Bookmark name..." /></div>
                    <div>Url:  <input type="text" value={this.state.url} onChange={this.handleUrlChange} placeholder="url..." /></div>
                    <button value="Add" style={this.buttonStyle()}>{addOrEdditAcceptButtonText}</button>
                </form>
            </div>);
        }
        return <div style={buttonStyle} onClick={this.toggleShow}>{addOrEditToggleButtonText}</div>;
    }
}

// **AddOrEditTagComponent**
type AddTagProps = {
    isRoot?: boolean;
    tagToEdit?: TagModelState;
    onAddOrEdit: (newTag: AddOrEditTagResult) => void;
};
type AddTagState = {
    isFormVisible: boolean;
    name: string;
};

export class AddOrEditTag extends React.Component<AddTagProps, AddTagState> {
    constructor(props) {
        super(props);
        const tagToEdit = this.props.tagToEdit || '';
        this.state = { isFormVisible: false, name: tagToEdit && tagToEdit.name };
    }
    hasValue() {
        return this.state.name != null;
    }
    handleSubmit = (evt: React.FormEvent) => {
        evt.preventDefault();
        const { name } = this.state;
        const tagToEdit = this.props.tagToEdit;
        if (name) {
            if (tagToEdit) {
                const editTagResult: AddOrEditTagResult = { ...tagToEdit, oldName: tagToEdit.name };
                editTagResult.name = name;
                this.props.onAddOrEdit(editTagResult);
            }
            else {
                this.props.onAddOrEdit({ name: name, bookmarks: [] });
            }
            this.setState({ isFormVisible: false, name: '' });
        }
    };
    handleNameChange = (evt: React.ChangeEvent<HTMLInputElement>) => {
        this.setState({ name: evt.target.value });
    };
    toggleShow = () => {
        this.setState({ isFormVisible: !this.state.isFormVisible });
    };
    buttonStyle(): React.CSSProperties {
        if (!this.hasValue()) {
            return { cursor: 'not-allowed' };
        }
        return null;
    }
    render() {
        const buttonStyle = { display: 'inline-block', textDecoration: 'underline', cursor: 'pointer' };
        const { isRoot, tagToEdit } = this.props;
        const toggleAddButtonText = isRoot ? '(Add Tags)' : tagToEdit ? '(Edit Tag)' : '(Add SubTags)';
        const acceptText = tagToEdit ? 'Edit' : 'Add';
        if (this.state.isFormVisible) {
            return (<div>
                <div style={buttonStyle} onClick={this.toggleShow}>Close</div>
                <form onSubmit={this.handleSubmit}>
                    <div>Name: <input type="text" value={this.state.name} onChange={this.handleNameChange} placeholder="Tag name..." /></div>
                    <button value="Add" style={this.buttonStyle()}>{acceptText}</button>
                </form>
            </div>);
        }
        return <div style={buttonStyle} onClick={this.toggleShow}>{toggleAddButtonText}</div>;
    }
}

// **TagComponent**

type TagDispatchProps = {
    addBookmark: (bookmarkModel: BookmarkModel, tagName: string) => void;
    addOrEditTag: (addOrEditTagResult: AddOrEditTagResult, parentTagName: string) => void;
    onEditBookmark: (tag: TagModelState, bookmark: EditBookmark) => void;
}

type TagStateToProps = {
    parentTag: TagModelState;
    tag: TagModelState;
    subTags?: TagModelState[];
    bookmarks: BookmarkModel[];
}

type TagProps = TagStateToProps & TagDispatchProps;

export function Tag(props: TagProps) {
    const { tag, bookmarks, parentTag } = props;
    if (tag.hidden) {
        return null;
    }
    return (<li className="tag">
        <fieldset>
            <legend>Tag: {tag.name}</legend>
            <div>Bookmarks:</div>
            {<Bookmarks onEdit={bookmark => props.onEditBookmark(tag, bookmark)} bookmarks={props.bookmarks} />}

            {tag.subTags &&
                <>
                    <div>SubTags:</div>
                    <Tags tags={props.subTags} parentTag={tag} />
                </>}
            <AddOrEditBookmark onAddOrEdit={(newBookmark) => props.addBookmark(newBookmark, tag.name)} />
            <AddOrEditTag key={`add${tag.name}`} onAddOrEdit={(newTag) => props.addOrEditTag(newTag, parentTag.name)} />
            <AddOrEditTag key={`edit{tag.name}`} tagToEdit={tag} onAddOrEdit={(newTag) => props.addOrEditTag(newTag, parentTag && parentTag.name)} />
        </fieldset>
    </li>);
}

const bookmarksById = (bookmarkIds: string[], allBookmarks: BookmarkModel[]) => {
    return allBookmarks.filter(bookmark => bookmarkIds.some(x => x == bookmark.name));
}

const tagByName = (tagName: string, tags: TagModelState[]) => {
    return tags.find(x => x.name == tagName);
}

const tagMapStateToProps = (state: BookmarksAppState, ownProps: ConnectedTagProps): TagStateToProps => {
    const tag = tagByName(ownProps.tagName, state.tags);
    return ({
        tag: tag,
        subTags: tag.subTags && state.tags.filter(x => tag.subTags.some(tagName => tagName == x.name)),
        bookmarks: bookmarksById(tag.bookmarks, state.bookmarks),
        parentTag: tagByName(ownProps.parentTagName, state.tags)
    })
};
const tagMapDispatchToProps = (dispatch: React.Dispatch<Action<any>>): TagDispatchProps => {
    return {
        addBookmark: (bookmarkModel: BookmarkModel, tagName: string) => dispatch(addBookmark({ bookmarkModel: bookmarkModel, tagName: tagName })),
        addOrEditTag: (addOrEditTagResult: AddOrEditTagResult, parentTagName: string) => dispatch(addOrEditTag({ addOrEditTagResult: addOrEditTagResult, parentTagName: parentTagName })),
        onEditBookmark: (tag: TagModelState, bookmark: EditBookmark) => dispatch(editBookmark(bookmark))
    }
}

type ConnectedTagProps = {
    parentTagName: string;
    tagName: string;
}

export const ConnectedTag: React.ComponentClass<ConnectedTagProps> = connect(tagMapStateToProps, tagMapDispatchToProps)(Tag);

// **TagsComponent**

type TagsProps = {
    parentTag: TagModelState;
    tags: TagModelState[];
};

export function Tags(props: TagsProps) {
    const { tags, parentTag } = props;
    if (tags) {
        const parentTagName = parentTag && parentTag.name;
        const tagItems = tags.map(tag => <ConnectedTag key={tag.name} parentTagName={parentTagName} tagName={tag.name} />);
        return (<ul style={{ listStyleType: 'none', paddingInlineStart: 0 }}>{tagItems}</ul>);
    }
    return null;
}

// **TagSearchComponent**

type TagSearchProps = {
    searachText: string;
    onSearchChange: (value: string) => void;
};

export function TagSearch(props: TagSearchProps) {
    const { searachText, onSearchChange } = props;
    return (<div><input type="text" placeholder="Search tags..." value={searachText} onChange={(evt) => onSearchChange(evt.target.value)} /></div>);
}

// **TagsRootComponent**

export type TagsRootDispatchProps = {
    addOrEditTag: (addOrEditTagResult: AddOrEditTagResult) => void;
    search: (searchValue: string) => void;
}

export type TagsRootStateProps = {
    tags: TagModelState[];
    searachText?: string;
}

export type TagsRootProps = TagsRootStateProps & TagsRootDispatchProps;

export function TagsRoot(props: TagsRootProps) {
    const { tags, search, searachText, addOrEditTag } = props;

    if (tags) {
        return (<div>
            <TagSearch searachText={searachText} onSearchChange={search} />
            <Tags tags={tags} parentTag={null} />
            <AddOrEditTag isRoot={true} onAddOrEdit={(newTag) => addOrEditTag(newTag)} />
        </div>);
    }
    return null;
}

const tagsRootMapStateToProps = (state: BookmarksAppState): TagsRootStateProps => {
    return {
        tags: state.tags.filter(x => x.isRoot),
        searachText: ''
    };
}

const tagsRootMapDispatchToProps = (dispatch: React.Dispatch<Action<any>>): TagsRootDispatchProps => {
    return {
        addOrEditTag: (addOrEditTagResult: AddOrEditTagResult) => dispatch(addOrEditTag({ addOrEditTagResult: addOrEditTagResult, parentTagName: null })),
        search: (searchValue: string) => dispatch(search({ searchValue: searchValue }))
    }
}

export const ConnectedTagsRoot = connect(tagsRootMapStateToProps, tagsRootMapDispatchToProps)(TagsRoot);