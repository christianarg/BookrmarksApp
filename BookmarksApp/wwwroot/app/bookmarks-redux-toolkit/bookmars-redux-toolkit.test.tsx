import * as React from 'react';
import { shallow, render } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { TagsRoot } from '../bookmarks-redux/bookmarks-redux';
import { TagModelState } from './bookmarks-redux-toolkit';


describe("TagsRoot tests", () => {

    test("Render root tag  empty tags", () => {
        // ARRANGE
        const tags: TagModelState[] = []
        // ACT
        const tagsRoot = shallow(<TagsRoot tags={tags} addOrEditTag={null} search={null} />);

        // ASSERT
        expect(shallowToJson(tagsRoot)).toMatchSnapshot();
    });

    test("Render root tag with tags", () => {
        // ARRANGE
        const tags: TagModelState[] = [{
            name: 'tag',
            bookmarks: ['bookmark1','bookmark2']
        }]
        // ACT
        const tagsRoot = shallow(<TagsRoot tags={tags} addOrEditTag={null} search={null} />);

        // ASSERT
        expect(shallowToJson(tagsRoot)).toMatchSnapshot();
    });

    test("Add root tag", () => {
        // ARRANGE
        // TODO: me dapalo
        //const tagRoot = render(<TagsRoot />);
        //tagRoot.find("input")
        // ACT

        // ASSERT
    });

    test("Remove root tag", () => {
        // ARRANGE

        // ACT

        // ASSERT
    });

    test("Edit root tag", () => {
        // ARRANGE

        // ACT

        // ASSERT
    });

    test("Render nested tag", () => {
        // ARRANGE

        // ACT

        // ASSERT
    });

});
