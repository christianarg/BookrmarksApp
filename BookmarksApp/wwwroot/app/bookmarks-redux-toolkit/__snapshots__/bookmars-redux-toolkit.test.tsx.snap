// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagsRoot tests Render root tag  empty tags 1`] = `
<div>
  <TagSearch
    onSearchChange={null}
  />
  <Tags
    parentTag={null}
    tags={Array []}
  />
  <AddOrEditTag
    isRoot={true}
    onAddOrEdit={[Function]}
  />
</div>
`;

exports[`TagsRoot tests Render root tag with tags 1`] = `
<div>
  <TagSearch
    onSearchChange={null}
  />
  <Tags
    parentTag={null}
    tags={
      Array [
        Object {
          "bookmarks": Array [
            "bookmark1",
            "bookmark2",
          ],
          "name": "tag",
        },
      ]
    }
  />
  <AddOrEditTag
    isRoot={true}
    onAddOrEdit={[Function]}
  />
</div>
`;
